### PR TITLE
Cache last written LSN for last  updated relations to reduce wait LSN  time for queries to other relations

### PR DESCRIPTION
**Pull request by knizhnik**
See
https://github.com/neondatabase/neon/pull/1763
https://github.com/neondatabase/neon/pull/1793

### CODE REVIEW

File: src/backend/access/transam/xlog.c [hunk 1]
```diff
@@ -8814,11 +8828,31 @@ GetInsertRecPtr(void)
...
+			}
+		}
+	}
+	else
+	{
+		/* Find maximum of all cached LSNs */
+		for (int i = 0; i < LAST_WRITTEN_CACHE_SIZE; i++)
+		{
+			if (XLogCtl->lastWrittenPageCache[i].lsn > lsn)
+				lsn = XLogCtl->lastWrittenPageCache[i].lsn;
```
Comment (lubennikovaav) [inline src/backend/access/transam/xlog.c:8853]: Why not max(XLogCtl->lastWrittenPageLSN, max_cached) ?
This logic deserves some comment.

**Comment by lubennikovaav** [comment]
Looks cool. Have you tested how it affects the performance?

**Comment by knizhnik** [comment]
> Looks cool. Have you tested how it affects the performance?

Yes, I am able now to login to database and perform read-only queries during `pgbench -i -s 1000` with default 100MB `max_wal_replication_lag` and old backpressure algorithm.

**Comment by hlinnaka** [comment]
I'm a little worried about doing so much while holding the info_lck spinlock. It's actually pretty random that we're using info_lck to protect the lastWrittenPageLsn in the first place. Perhaps it's time to move that out of the XLogCtl struct and have a separate lock for it.

But yeah, I agree with the general idea.

One similar idea I had is to have e.g. 100 last-written-LSN values in a hash table. When you evict a page, you calculate a hash of the BufferTag of the page you're evicting, and store the LSN at that slot in the hash table. When reading, similarly calculate the hash of the BufferTag, and use the LSN at that slot. There would be some "false sharing", but that's OK. However, it wouldn't be so efficient for the specific problem you're addressing. If you do a large COPY, it would very quickly update all the LSNs in the hash table, as it modifies a lot of pages.

**Comment by knizhnik** [comment]
Checking just four element is not a huge work. But I understand your concerns about spinlock. Unfortunately I do not see better alternative: using rw-lock may be even more expensive.

I also thought about replacing RelFileNode with just relation ID. The probability of collision is highly unlikely and comparing just one integer instead of three may have some positive impact on performance.

Concerning your suggestion with hash table - I am not sure which access pattern is can optimize? Random updates (like in pgbench)?
But in this case probability that entry will not be flushed from the cache before been accessed next time will be close to 0.
It can work only if we perform updates of very small working set (~100 pages). But in this case this 100  pages will always present in shared buffer and so no need to track their LSNs in any cache. Also such shared cache can easily become a bottleneck.

Yes, in this PR I mostly address bulk loading data in the database. When just one table is populated with data (may be with indexes). When there are many indexes or several tables are simultaneously populated, then such small cache (4) will not help. 

I think that we definitely need some kind of wait_lsn optimization. The first thing which people will do with Neon is try to upload some data to it. But as far as waiting end of upload is boring, then they most likely will want to establish one more connection and check (using pg_stat_activity or whatever else) how things are going. And if them  are just not be able to login - it will have very negative impression. More conservative backpressure settings may help. But I prefer to combine both mechanisms.

**Comment by hlinnaka** [comment]
> Concerning your suggestion with hash table - I am not sure which access pattern is can optimize? Random updates (like in pgbench)? But in this case probability that entry will not be flushed from the cache before been accessed next time will be close to 0. It can work only if we perform updates of very small working set (~100 pages). But in this case this 100 pages will always present in shared buffer and so no need to track their LSNs in any cache. Also such shared cache can easily become a bottleneck.

Agreed, a small hash table would not be very helpful. Maybe if the hash table was much larger, say, 1000000 entries. But yeah, this would need some more thinking.

> I think that we definitely need some kind of wait_lsn optimization. The first thing which people will do with Neon is try to upload some data to it. But as far as waiting end of upload is boring, then they most likely will want to establish one more connection and check (using pg_stat_activity or whatever else) how things are going. And if them are just not be able to login - it will have very negative impression. More conservative backpressure settings may help. But I prefer to combine both mechanisms.

+1

**Comment by lubennikovaav** [comment]
`pgbench -i -s 1` fails with assertion on this branch:

```
TRAP: FailedAssertion("lsn != InvalidXLogRecPtr", File: "/home/anastasia/zenith/tmp_install/build/../../vendor/postgres/contrib/neon/pagestore_smgr.c", Line: 635, PID: 219368)
```

File: src/backend/access/transam/xlog.c [hunk 2]
```diff
@@ -607,6 +610,11 @@ typedef struct XLogCtlInsert
 	WALInsertLockPadded *WALInsertLocks;
 } XLogCtlInsert;
 
+typedef struct RnodeForkKey {
+	Oid rnode;
```
Comment (lubennikovaav) [inline src/backend/access/transam/xlog.c:614]: let's add a comment about why this is safe to use only `rnode+forknum` as a key

File: contrib/neon/pagestore_smgr.c [hunk 3]
```diff
@@ -1345,7 +1345,7 @@ zenith_dbsize(Oid dbNode)
 	XLogRecPtr request_lsn;
 	bool		latest;
 
-	request_lsn = zenith_get_request_lsn(&latest);
+	request_lsn = zenith_get_request_lsn(&latest, NULL, InvalidForkNumber);
```
Comment (lubennikovaav) [inline contrib/neon/pagestore_smgr.c:1348]: `zenith_get_request_lsn` is not ready to accept NULL argument:

``` GetLastWrittenPageLSN(rnode->relNode```

File: src/backend/access/transam/xlog.c [hunk 4]
```diff
@@ -8828,11 +8872,41 @@ GetLastWrittenPageLSN(void)
  * SetLastWrittenPageLSN -- Set maximal LSN of written page
```
Comment (lubennikovaav) [inline src/backend/access/transam/xlog.c:8872]: please expand this comment to mention caching

**Comment by knizhnik** [comment]
> I think we must also use forknum as a part of the key.

Sorry, this fix is not correct.
This cache should not be precise as far as it stores maximum LSN.
This is why it is not possible to take in account forknumber, dbnode, spcnode.
Even if there are several relation matching cache entry, we still maintain max(lsn) for them.
And there is no difference with single lastWrittenPageLsn  we have before.

Let me investigate why `pgench -i -s1`  is failed.May be we should check for valid LSN in SetLastWrittenPageLsn()
But in any case you fix is not compatible with my last optimization.
Either key should uniquely idetify relation, and so contain all four components: spcnode, dbnode, relnode, forknum
Either is can be partial key and so forknumer is not needed as well.

**Comment by lubennikovaav** [comment]
> But in any case you fix is not compatible with my last optimization. 

I'm not sure, if I understand the problem. My "fix" just adds an unnecessary part to the key or it is incompatible in some other way?

**Comment by knizhnik** [comment]
> My "fix" just adds an unnecessary part to the key or it is incompatible in some other way?

Relation is identified using four components: spcnode, dbnode, relnode, forknum
Originally I used only first three of them (RelFileNode) as a key. So there is no one-to-one mapping 
between key and relations.
Then I decided to use only relnode as a key: probability of collision is assumed to be very small and non-critical.
Then you add forknum. But (relnode,forknum) is not unique key! That is why I said that you fix has no sense.

The reason of the problem is that we sometimes pass lsn=0 in SetLastWrittenLsn.
Before his PR it doesn't cause any problem because we in any case assign maximal value.
But now we can associate InvalidXLogRecPtr with this relation and then GetLastValidLsn may return it which cause assertion failure in `zenith_get_request_lsn`. But is enough just to ad check for InvalidXLogRecPtr in SetLastWritenPageLsn.

**Comment by aome510** [comment]
> One similar idea I had is to have e.g. 100 last-written-LSN values in a hash table. When you evict a page, you calculate a hash of the BufferTag of the page you're evicting, and store the LSN at that slot in the hash table. When reading, similarly calculate the hash of the BufferTag, and use the LSN at that slot. There would be some "false sharing", but that's OK. However, it wouldn't be so efficient for the specific problem you're addressing. If you do a large COPY, it would very quickly update all the LSNs in the hash table, as it modifies a lot of pages.

@hlinnaka idea is similar to what [Socrates](https://www.microsoft.com/en-us/research/uploads/prod/2019/05/socrates.pdf) does to estimate the latest LSN for a page. Quote from section 4.4:

> Ideally, X-LSN would be the most recent page LSN for Page X. However, the Primary cannot remember the LSNs of all pages it has evicted (essentially the whole database). Instead, the Primary builds a hash map (on pageId) which stores in each bucket the highest LSN for every page evicted from the Primary keyed by pageId. Given that Page X was evicted at some point from the Primary, this mechanism will guarantee to give an X-LSN value that is at least as large as the largest LSN for Page X and is, thus, safe.

It's true that this approach may not work well with bulk loading if using a small bucket size for the hashmap. We can probably tweak the hash map to key by `relation_id` instead of `page_id`, so it is more suitable for the bulk-loading task.

IMHO, for bulk-loading task, using `page_id` may be not that bad because the hash map only stores LSN of evicted pages. If the shared buffer is large enough, the difference between the maximum LSN of an evicted pages and the last record LSN processed by the page server should be close, resulting in a small wait time.

**Comment by knizhnik** [comment]
> @hlinnaka idea is similar to what [Socrates](https://www.microsoft.com/en-us/research/uploads/prod/2019/05/socrates.pdf) does to estimate the latest LSN for a page.

My main concern is synchronzation needed to access this shared cache.
Certainly, we can use standard postgres approach with hash partitioning (when each partitioned is synchronizae its own RW-lock). But I am not sure that it will completely eliminate the probem.

Also I think that using PageId is key is not so good idea. In case of random updates (like in pgbench), there will e a lot of cache collisions (i.e. all baskets will have recent LSNs). May be it is better to split relation into larger chunks (i.e. 1Mb or even more) and maintain latest LSN for each chunk? It also will not help much in case of random updates. But will be more efficient in case of appending data to multiple relations.

@hlinnaka - what do you think? Should I try to implement such more sophisticated caching of last recent LSNs?
Or current solution with very small cache with `relation_id` as a key is enough for now?

**Comment by hlinnaka** [comment]
> > @hlinnaka idea is similar to what [Socrates](https://www.microsoft.com/en-us/research/uploads/prod/2019/05/socrates.pdf) does to estimate the latest LSN for a page.
> 
> My main concern is synchronzation needed to access this shared cache. Certainly, we can use standard postgres approach with hash partitioning (when each partitioned is synchronizae its own RW-lock). But I am not sure that it will completely eliminate the probem.

I think that's good enough. The synchronization overhead is tiny compared to sending a request over the network.

> Also I think that using PageId is key is not so good idea. In case of random updates (like in pgbench), there will e a lot of cache collisions (i.e. all baskets will have recent LSNs). May be it is better to split relation into larger chunks (i.e. 1Mb or even more) and maintain latest LSN for each chunk? It also will not help much in case of random updates. But will be more efficient in case of appending data to multiple relations.

Yeah, I think that's worth exploring.

If we have a more fine-grained cache with page-ids or chunks, can we use it for smgrnblocks() and smgrexists() calls too? We also have a relation size cache, so maybe 

> @hlinnaka - what do you think? Should I try to implement such more sophisticated caching of last recent LSNs? Or current solution with very small cache with `relation_id` as a key is enough for now?

The small per-relation cache certainly helps, but it's very easy to overflow it...

**Comment by aome510** [comment]
So I tried to run this patch with the new WAL backpressure tests

Results:

latest `main` (https://app.circleci.com/pipelines/github/neondatabase/neon/7285/workflows/e18911a4-4cf7-43ee-863b-2d6ec11540db/jobs/74648):

<details>

```
test_heavy_write_workload[vanilla-10-5-5].run_duration: 37.701 s
test_heavy_write_workload[vanilla-10-5-5].read_latency_max: 0.021 s
test_heavy_write_workload[vanilla-10-5-5].read_latency_avg: 0.001 s
test_heavy_write_workload[vanilla-10-5-5].read_latency_stdev: 0.004 s
test_heavy_write_workload[neon_off-10-5-5].run_duration: 227.946 s
test_heavy_write_workload[neon_off-10-5-5].lsn_write_lag_max: 4 MB
test_heavy_write_workload[neon_off-10-5-5].lsn_write_lag_avg: 1 MB
test_heavy_write_workload[neon_off-10-5-5].lsn_write_lag_stdev: 1 MB
test_heavy_write_workload[neon_off-10-5-5].read_latency_max: 0.206 s
test_heavy_write_workload[neon_off-10-5-5].read_latency_avg: 0.037 s
test_heavy_write_workload[neon_off-10-5-5].read_latency_stdev: 0.030 s
test_heavy_write_workload[neon_on-10-5-5].run_duration: 250.031 s
test_heavy_write_workload[neon_on-10-5-5].lsn_write_lag_max: 4 MB
test_heavy_write_workload[neon_on-10-5-5].lsn_write_lag_avg: 1 MB
test_heavy_write_workload[neon_on-10-5-5].lsn_write_lag_stdev: 1 MB
test_heavy_write_workload[neon_on-10-5-5].read_latency_max: 0.500 s
test_heavy_write_workload[neon_on-10-5-5].read_latency_avg: 0.048 s
test_heavy_write_workload[neon_on-10-5-5].read_latency_stdev: 0.063 s
test_pgbench_simple_update_workload[vanilla-45-100].run_duration: 45.447 s
test_pgbench_simple_update_workload[vanilla-45-100].read_latency_max: 0.057 s
test_pgbench_simple_update_workload[vanilla-45-100].read_latency_avg: 0.002 s
test_pgbench_simple_update_workload[vanilla-45-100].read_latency_stdev: 0.010 s
test_pgbench_simple_update_workload[neon_off-45-100].run_duration: 46.292 s
test_pgbench_simple_update_workload[neon_off-45-100].lsn_write_lag_max: 0 MB
test_pgbench_simple_update_workload[neon_off-45-100].lsn_write_lag_avg: 0 MB
test_pgbench_simple_update_workload[neon_off-45-100].lsn_write_lag_stdev: 0 MB
test_pgbench_simple_update_workload[neon_off-45-100].read_latency_max: 0.040 s
test_pgbench_simple_update_workload[neon_off-45-100].read_latency_avg: 0.004 s
test_pgbench_simple_update_workload[neon_off-45-100].read_latency_stdev: 0.007 s
test_pgbench_simple_update_workload[neon_on-45-100].run_duration: 53.027 s
test_pgbench_simple_update_workload[neon_on-45-100].lsn_write_lag_max: 1 MB
test_pgbench_simple_update_workload[neon_on-45-100].lsn_write_lag_avg: 0 MB
test_pgbench_simple_update_workload[neon_on-45-100].lsn_write_lag_stdev: 0 MB
test_pgbench_simple_update_workload[neon_on-45-100].read_latency_max: 0.059 s
test_pgbench_simple_update_workload[neon_on-45-100].read_latency_avg: 0.005 s
test_pgbench_simple_update_workload[neon_on-45-100].read_latency_stdev: 0.009 s
test_pgbench_intensive_init_workload[vanilla-1000].run_duration: 128.655 s
test_pgbench_intensive_init_workload[vanilla-1000].read_latency_max: 0.017 s
test_pgbench_intensive_init_workload[vanilla-1000].read_latency_avg: 0.008 s
test_pgbench_intensive_init_workload[vanilla-1000].read_latency_stdev: 0.003 s
test_pgbench_intensive_init_workload[neon_off-1000].run_duration: 59.479 s
test_pgbench_intensive_init_workload[neon_off-1000].lsn_write_lag_max: 9 MB
test_pgbench_intensive_init_workload[neon_off-1000].lsn_write_lag_avg: 2 MB
test_pgbench_intensive_init_workload[neon_off-1000].lsn_write_lag_stdev: 2 MB
test_pgbench_intensive_init_workload[neon_off-1000].read_latency_max: 2.030 s
test_pgbench_intensive_init_workload[neon_off-1000].read_latency_avg: 0.775 s
test_pgbench_intensive_init_workload[neon_off-1000].read_latency_stdev: 0.477 s
test_pgbench_intensive_init_workload[neon_on-1000].run_duration: 141.701 s
test_pgbench_intensive_init_workload[neon_on-1000].lsn_write_lag_max: 15 MB
test_pgbench_intensive_init_workload[neon_on-1000].lsn_write_lag_avg: 13 MB
test_pgbench_intensive_init_workload[neon_on-1000].lsn_write_lag_stdev: 4 MB
test_pgbench_intensive_init_workload[neon_on-1000].read_latency_max: 27.029 s
test_pgbench_intensive_init_workload[neon_on-1000].read_latency_avg: 19.359 s
test_pgbench_intensive_init_workload[neon_on-1000].read_latency_stdev: 9.079 s
```

</details>

this patch + @hlinnaka's https://github.com/neondatabase/postgres/pull/175 (https://app.circleci.com/pipelines/github/neondatabase/neon/7286/workflows/9cba7367-a9eb-4a26-82bc-844e0bb7e9e5/jobs/74656):

<details>

```
test_heavy_write_workload[vanilla-10-5-5].run_duration: 39.358 s
test_heavy_write_workload[vanilla-10-5-5].read_latency_max: 0.006 s
test_heavy_write_workload[vanilla-10-5-5].read_latency_avg: 0.001 s
test_heavy_write_workload[vanilla-10-5-5].read_latency_stdev: 0.001 s
test_heavy_write_workload[neon_off-10-5-5].run_duration: 173.292 s
test_heavy_write_workload[neon_off-10-5-5].read_latency_max: 1.172 s
test_heavy_write_workload[neon_off-10-5-5].read_latency_avg: 0.401 s
test_heavy_write_workload[neon_off-10-5-5].read_latency_stdev: 0.341 s
test_heavy_write_workload[neon_off-10-5-5].lsn_write_lag_max: 17 MB
test_heavy_write_workload[neon_off-10-5-5].lsn_write_lag_avg: 15 MB
test_heavy_write_workload[neon_off-10-5-5].lsn_write_lag_stdev: 3 MB
test_heavy_write_workload[neon_on-10-5-5].run_duration: 188.862 s
test_heavy_write_workload[neon_on-10-5-5].lsn_write_lag_max: 17 MB
test_heavy_write_workload[neon_on-10-5-5].lsn_write_lag_avg: 14 MB
test_heavy_write_workload[neon_on-10-5-5].lsn_write_lag_stdev: 3 MB
test_heavy_write_workload[neon_on-10-5-5].read_latency_max: 1.304 s
test_heavy_write_workload[neon_on-10-5-5].read_latency_avg: 0.540 s
test_heavy_write_workload[neon_on-10-5-5].read_latency_stdev: 0.348 s
test_pgbench_simple_update_workload[vanilla-45-100].run_duration: 45.135 s
test_pgbench_simple_update_workload[vanilla-45-100].read_latency_max: 0.050 s
test_pgbench_simple_update_workload[vanilla-45-100].read_latency_avg: 0.002 s
test_pgbench_simple_update_workload[vanilla-45-100].read_latency_stdev: 0.009 s
test_pgbench_simple_update_workload[neon_off-45-100].run_duration: 46.109 s
test_pgbench_simple_update_workload[neon_off-45-100].read_latency_max: 0.027 s
test_pgbench_simple_update_workload[neon_off-45-100].read_latency_avg: 0.003 s
test_pgbench_simple_update_workload[neon_off-45-100].read_latency_stdev: 0.005 s
test_pgbench_simple_update_workload[neon_off-45-100].lsn_write_lag_max: 1 MB
test_pgbench_simple_update_workload[neon_off-45-100].lsn_write_lag_avg: 0 MB
test_pgbench_simple_update_workload[neon_off-45-100].lsn_write_lag_stdev: 0 MB
test_pgbench_simple_update_workload[neon_on-45-100].run_duration: 45.501 s
test_pgbench_simple_update_workload[neon_on-45-100].lsn_write_lag_max: 1 MB
test_pgbench_simple_update_workload[neon_on-45-100].lsn_write_lag_avg: 0 MB
test_pgbench_simple_update_workload[neon_on-45-100].lsn_write_lag_stdev: 0 MB
test_pgbench_simple_update_workload[neon_on-45-100].read_latency_max: 0.065 s
test_pgbench_simple_update_workload[neon_on-45-100].read_latency_avg: 0.005 s
test_pgbench_simple_update_workload[neon_on-45-100].read_latency_stdev: 0.011 s
test_pgbench_intensive_init_workload[vanilla-1000].run_duration: 99.138 s
test_pgbench_intensive_init_workload[vanilla-1000].read_latency_max: 0.013 s
test_pgbench_intensive_init_workload[vanilla-1000].read_latency_avg: 0.007 s
test_pgbench_intensive_init_workload[vanilla-1000].read_latency_stdev: 0.001 s
test_pgbench_intensive_init_workload[neon_off-1000].run_duration: 56.939 s
test_pgbench_intensive_init_workload[neon_off-1000].read_latency_max: 0.763 s
test_pgbench_intensive_init_workload[neon_off-1000].read_latency_avg: 0.149 s
test_pgbench_intensive_init_workload[neon_off-1000].read_latency_stdev: 0.136 s
test_pgbench_intensive_init_workload[neon_off-1000].lsn_write_lag_max: 6 MB
test_pgbench_intensive_init_workload[neon_off-1000].lsn_write_lag_avg: 2 MB
test_pgbench_intensive_init_workload[neon_off-1000].lsn_write_lag_stdev: 1 MB
test_pgbench_intensive_init_workload[neon_on-1000].run_duration: 87.980 s
test_pgbench_intensive_init_workload[neon_on-1000].read_latency_max: 1.153 s
test_pgbench_intensive_init_workload[neon_on-1000].read_latency_avg: 0.210 s
test_pgbench_intensive_init_workload[neon_on-1000].read_latency_stdev: 0.282 s
test_pgbench_intensive_init_workload[neon_on-1000].lsn_write_lag_max: 15 MB
test_pgbench_intensive_init_workload[neon_on-1000].lsn_write_lag_avg: 10 MB
test_pgbench_intensive_init_workload[neon_on-1000].lsn_write_lag_stdev: 4 MB
```

</details>

Differences (baseline is this patch + Heikki's patch):
```
key: test_heavy_write_workload[neon_off-10-5-5].lsn_write_lag_avg, diff: 14.00, diff_percentage: 1400.00%
key: test_heavy_write_workload[neon_off-10-5-5].lsn_write_lag_max, diff: 13.00, diff_percentage: 325.00%
key: test_heavy_write_workload[neon_off-10-5-5].lsn_write_lag_stdev, diff: 2.00, diff_percentage: 200.00%
key: test_heavy_write_workload[neon_off-10-5-5].read_latency_avg, diff: 0.36, diff_percentage: 983.78%
key: test_heavy_write_workload[neon_off-10-5-5].read_latency_max, diff: 0.97, diff_percentage: 468.93%
key: test_heavy_write_workload[neon_off-10-5-5].read_latency_stdev, diff: 0.31, diff_percentage: 1036.67%
key: test_heavy_write_workload[neon_off-10-5-5].run_duration, diff: -54.65, diff_percentage: -23.98%
key: test_heavy_write_workload[neon_on-10-5-5].lsn_write_lag_avg, diff: 13.00, diff_percentage: 1300.00%
key: test_heavy_write_workload[neon_on-10-5-5].lsn_write_lag_max, diff: 13.00, diff_percentage: 325.00%
key: test_heavy_write_workload[neon_on-10-5-5].lsn_write_lag_stdev, diff: 2.00, diff_percentage: 200.00%
key: test_heavy_write_workload[neon_on-10-5-5].read_latency_avg, diff: 0.49, diff_percentage: 1025.00%
key: test_heavy_write_workload[neon_on-10-5-5].read_latency_max, diff: 0.80, diff_percentage: 160.80%
key: test_heavy_write_workload[neon_on-10-5-5].read_latency_stdev, diff: 0.28, diff_percentage: 452.38%
key: test_heavy_write_workload[neon_on-10-5-5].run_duration, diff: -61.17, diff_percentage: -24.46%
key: test_heavy_write_workload[vanilla-10-5-5].read_latency_avg, diff: 0.00, diff_percentage: 0.00%
key: test_heavy_write_workload[vanilla-10-5-5].read_latency_max, diff: -0.02, diff_percentage: -71.43%
key: test_heavy_write_workload[vanilla-10-5-5].read_latency_stdev, diff: -0.00, diff_percentage: -75.00%
key: test_heavy_write_workload[vanilla-10-5-5].run_duration, diff: 1.66, diff_percentage: 4.40%
key: test_pgbench_simple_update_workload[neon_off-45-100].read_latency_avg, diff: -0.00, diff_percentage: -25.00%
key: test_pgbench_simple_update_workload[neon_off-45-100].read_latency_max, diff: -0.01, diff_percentage: -32.50%
key: test_pgbench_simple_update_workload[neon_off-45-100].read_latency_stdev, diff: -0.00, diff_percentage: -28.57%
key: test_pgbench_simple_update_workload[neon_off-45-100].run_duration, diff: -0.18, diff_percentage: -0.40%
key: test_pgbench_simple_update_workload[neon_on-45-100].lsn_write_lag_max, diff: 0.00, diff_percentage: 0.00%
key: test_pgbench_simple_update_workload[neon_on-45-100].read_latency_avg, diff: 0.00, diff_percentage: 0.00%
key: test_pgbench_simple_update_workload[neon_on-45-100].read_latency_max, diff: 0.01, diff_percentage: 10.17%
key: test_pgbench_simple_update_workload[neon_on-45-100].read_latency_stdev, diff: 0.00, diff_percentage: 22.22%
key: test_pgbench_simple_update_workload[neon_on-45-100].run_duration, diff: -7.53, diff_percentage: -14.19%
key: test_pgbench_simple_update_workload[vanilla-45-100].read_latency_avg, diff: 0.00, diff_percentage: 0.00%
key: test_pgbench_simple_update_workload[vanilla-45-100].read_latency_max, diff: -0.01, diff_percentage: -12.28%
key: test_pgbench_simple_update_workload[vanilla-45-100].read_latency_stdev, diff: -0.00, diff_percentage: -10.00%
key: test_pgbench_simple_update_workload[vanilla-45-100].run_duration, diff: -0.31, diff_percentage: -0.69%
key: test_pgbench_intensive_init_workload[neon_off-1000].lsn_write_lag_avg, diff: 0.00, diff_percentage: 0.00%
key: test_pgbench_intensive_init_workload[neon_off-1000].lsn_write_lag_max, diff: -3.00, diff_percentage: -33.33%
key: test_pgbench_intensive_init_workload[neon_off-1000].lsn_write_lag_stdev, diff: -1.00, diff_percentage: -50.00%
key: test_pgbench_intensive_init_workload[neon_off-1000].read_latency_avg, diff: -0.63, diff_percentage: -80.77%
key: test_pgbench_intensive_init_workload[neon_off-1000].read_latency_max, diff: -1.27, diff_percentage: -62.41%
key: test_pgbench_intensive_init_workload[neon_off-1000].read_latency_stdev, diff: -0.34, diff_percentage: -71.49%
key: test_pgbench_intensive_init_workload[neon_off-1000].run_duration, diff: -2.54, diff_percentage: -4.27%
key: test_pgbench_intensive_init_workload[neon_on-1000].lsn_write_lag_avg, diff: -3.00, diff_percentage: -23.08%
key: test_pgbench_intensive_init_workload[neon_on-1000].lsn_write_lag_max, diff: 0.00, diff_percentage: 0.00%
key: test_pgbench_intensive_init_workload[neon_on-1000].lsn_write_lag_stdev, diff: 0.00, diff_percentage: 0.00%
key: test_pgbench_intensive_init_workload[neon_on-1000].read_latency_avg, diff: -19.15, diff_percentage: -98.92%
key: test_pgbench_intensive_init_workload[neon_on-1000].read_latency_max, diff: -25.88, diff_percentage: -95.73%
key: test_pgbench_intensive_init_workload[neon_on-1000].read_latency_stdev, diff: -8.80, diff_percentage: -96.89%
key: test_pgbench_intensive_init_workload[neon_on-1000].run_duration, diff: -53.72, diff_percentage: -37.91%
key: test_pgbench_intensive_init_workload[vanilla-1000].read_latency_avg, diff: -0.00, diff_percentage: -12.50%
key: test_pgbench_intensive_init_workload[vanilla-1000].read_latency_max, diff: -0.00, diff_percentage: -23.53%
key: test_pgbench_intensive_init_workload[vanilla-1000].read_latency_stdev, diff: -0.00, diff_percentage: -66.67%
key: test_pgbench_intensive_init_workload[vanilla-1000].run_duration, diff: -29.52, diff_percentage: -22.94%
```

## Notes
- in general, at least based on the new tests, this patch significantly improves the read latency.
- notice that there is a big difference between the run duration of `test_pgbench_intensive_init_workload[neon_on-1000]`, `test_heavy_write_workload[neon-*]` tests, which is probably related to [Heikki's comment](https://github.com/neondatabase/neon/pull/1891#issuecomment-1151549148)
> With this workload, postgres will call smgrexists() on the FSM fork repeatedly, whenever it needs to extend the index. That again causes the backend to wait for the pageserver to catch up, draining any lag. (I noticed this by looking at pg_stat_activity, saw the backends blocked on the "Extension" wait event, and then I connected to the backend process with 'gdb' to see what exactly it was doing)
- in `test_pgbench_intensive_init_workload`, even though the read relation is never in the cache, it still lags behind the latest written LSN as LSNs of hot relations are stored in the cache, so the overall read latency is not that big.

File: src/backend/access/transam/xlog.c [hunk 5]
```diff
@@ -748,7 +751,15 @@ typedef struct XLogCtlData
 	 * XLOG_FPW_CHANGE record that instructs full_page_writes is disabled.
 	 */
 	XLogRecPtr	lastFpwDisableRecPtr;
-	XLogRecPtr  lastWrittenPageLSN;
+
+	/*
+	 * Cache of last written page LSN.
+	 * We store this value for up to LAST_WRITTEN_CACHE_SIZE relations + maximum for all other relations.
+	 */
+	XLogRecPtr	lastWrittenPageLsn;
```
Comment (aome510) [inline src/backend/access/transam/xlog.c:759]:
```suggestion
	XLogRecPtr	lastWrittenPageNonCacheLsn;
```

I know there is a comment described this above, but it's still better to make the variable self-explained. At first, I thought the non-cache relations will be default to use the latest written LSN based on the variable name.

**Comment by knizhnik** [comment]
> in test_pgbench_intensive_init_workload, even though the read relation is never in the cache, the query LSN still lags behind the latest written LSN as LSNs of hot relations are stored in the cache, so the overall read latency is not that big.

Please notice, that this LSN cache works not only only when relation is present in the cache (cache hit).
Almost opposite: this cache allows to increase speed of read-only queries which access other relations (which LSNs are not cached). Cache allows to postpone update of `lastWrittenPageNonCacheLsn` and so use old LSN for retrieving pages of this relations from server. Certainly it cn work only if number of updated relations is relatively small. But there can be arbitrary number of read-only relations.

**Comment by knizhnik** [comment]
> If we have a more fine-grained cache with page-ids or chunks, can we use it for smgrnblocks() and smgrexists() calls too? We also have a relation size cache, so maybe

Not sure that the same cache can be used both for caching relation size and keeping last written LSN.
First of all how we can locate the last page?  Just standard hash is not able to do it.
Also what will happen if cache size is limited and last page was evicted from the cache, but some other page - not?
It seems to be more useful and realistic to try to combine last written lsn cache and relation size cache

**Comment by knizhnik** [comment]
I have created another PR #177 with ls written LSN maintaine for each relation chunk (8Mb).

**Comment by knizhnik** [comment]
Replaced with #191

**Pull request closed without merging**